File: 1547-minimum-cost-to-cut-a-stick/1547-minimum-cost-to-cut-a-stick.java
class Solution {
    Map<String, Long> dp;
    final int INF = Integer.MAX_VALUE;
    public int minCost(int n, int[] cuts) {
        Arrays.sort(cuts);
        dp = new HashMap<>();
        return (int)solve(0, n, cuts);
    }
    long solve(int l, int r, int[] cuts) {
        if(l>=r) return 0;
        String key = l + "," + r;
        if(dp.containsKey(key)) return dp.get(key);
        long ans = INF;
        boolean flag = false;
        for(int cut : cuts) {
            if(cut > l && cut < r) {
                flag = true;
                long val = solve(l, cut, cuts) + solve(cut, r, cuts) + (r-l);
                ans = Math.min(ans, val);
            }
        }
        if(!flag) ans = 0;
        dp.put(key, ans);
        return ans;
    }
}

File: 1547-minimum-cost-to-cut-a-stick/1547-minimum-cost-to-cut-a-stick.js
/**
 * @param {number} n
 * @param {number[]} cuts
 * @return {number}
 */
var minCost = function(n, cuts) {
    cuts.sort((a, b) => a - b);
    let dp = new Map();

    function solve(l, r) {
        if(l >= r) return 0;
        let key = `${l},${r}`;
        if(dp.has(key)) return dp.get(key);

        let ans = Number.MAX_SAFE_INTEGER;

        for(let cut of cuts) {
            if(l < cut && r > cut) {
                let val = solve(l, cut) + solve(cut, r) + (r - l);
                ans = Math.min(ans, val);
            }
        }

        if(ans === Number.MAX_SAFE_INTEGER) {
            ans = 0;
        }

        dp.set(key, ans);

        return ans;
    }

    return solve(0, n);
};